provider/pkg/provider/container.go

```diff
@@ -0,0 +1,388 @@
+				logging.V(9).Infof("[%s] processing diff on field: %q", label, k)
+				var diffPropPath, propPath resource.PropertyPath
+				diffPropPath, err = resource.ParsePropertyPath(k)
+				if err != nil {
+					err = fmt.Errorf("failed to parse property path: %q: %w", k, err)
+					break
+				}
+				propPath, err = resource.ParsePropertyPath(key)
+				if err != nil {
+					err = fmt.Errorf("failed to parse property path: %q: %w", k, err)
+					break
+				}
+				if propPath.Contains(diffPropPath) {
+					logging.V(9).Infof("[%s] processing diff for %q with handler for %q",
+						label, diffPropPath.String(), propPath.String())
+					logging.V(9).Infof("[%s] waiting for nodepool to reach resting state", label)
+					_, err = waitForNodepoolRestingState(providerInstance, res, inputs, oldState)
+					if err != nil {
+						break
+					}
+					logging.V(9).Infof("[%s] calling update handler", label)
+					err = updateHandler(providerInstance, urn, label, res, inputs, oldState)
+					if err != nil {
+						logging.V(9).Infof("[%s] failure updating: %+v", label, err)
+						break
+					}
+				}
+			}
+			if err != nil {
+				break
+			}
+		}
+	}
+	uri, readErr := res.Read.Endpoint.URI(inputs.Mappable(), oldState.Mappable())
+	if readErr != nil {
+		return nil, readErr
+	}
+	resp, err2 := readNodepoolStatus(providerInstance, uri)
+	if err2 != nil {
+		logging.V(9).Infof("[%s] Failed to read nodepool status: %v", label, err2)
+		return nil, fmt.Errorf("failed to retrieve nodepool status: %v. Previous error: %w", err2, err)
+	}
+	return resp, err
+}
+
+func readNodepoolStatus(
+	p *googleCloudProvider,
+	uri string,
+) (map[string]interface{}, error) {
+	// Read the current state of the resource from the API.
+	newState, err := retryRequest(p.client, "GET", uri, "", nil)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching nodepool status: %w", err)
+	}
+	return newState, nil
+}
+
+func waitForNodepoolRestingState(p *googleCloudProvider,
+	res *resources.CloudAPIResource,
+	inputs,
+	oldState resource.PropertyMap,
+) (map[string]interface{}, error) {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 15*time.Minute)
+	defer cancelFunc()
+	retryPolicy := backoff.Backoff{
+		Min:    1 * time.Second,
+		Max:    15 * time.Second,
+		Factor: 1.5,
+		Jitter: true,
+	}
+
+	uri, err := res.Read.Endpoint.URI(inputs.Mappable(), oldState.Mappable())
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		resp, err := readNodepoolStatus(p, uri)
+		if err != nil {
+			return nil, err
+		}
+
+		status, hasStatus := resp["status"].(string)
+		if hasStatus && isNodepoolInRestingStatus(status) {
+			return resp, nil
+		}
+		select {
+		case <-time.After(retryPolicy.Duration()):
+		case <-ctx.Done():
+			return nil, fmt.Errorf("timeout polling for nodepool resting state: %q", uri)
+		}
+	}
+
+}
+
+func isNodepoolInRestingStatus(status string) bool {
+	switch status {
+	case "RUNNING", "RUNNING_WITH_ERROR", "ERROR":
+		return true
```

Comment:
I'd suggest adding a default case that prints/logs a warning for unhandled status values so that it's easier to catch changes here.

Reply:
There are multiple non-resting states so logging might be noisy. I have added a status log which should make it clear if we are stuck in an unaccounted-for terminal state much more readily.